server/src/main/java/org/elasticsearch/readiness/ReadinessService.java

```diff
@@ -0,0 +1,226 @@
+     * @return the bound address for the readiness service
+     */
+    public BoundTransportAddress boundAddress() {
+        InetSocketAddress boundAddress = boundSocket.get();
+        if (boundAddress == null) {
+            return null;
+        }
+        TransportAddress publishAddress = new TransportAddress(boundAddress);
+        return new BoundTransportAddress(new TransportAddress[] { publishAddress }, publishAddress);
+    }
+
+    ServerSocketChannel setupSocket() {
+        assert PORT.get(environment.settings()) >= 0;
+        InetAddress localhost = InetAddress.getLoopbackAddress();
+        int portNumber = PORT.get(environment.settings());
+
+        AccessController.doPrivileged((PrivilegedAction<Void>) () -> {
+            try {
+                serverChannel = ServerSocketChannel.open(StandardProtocolFamily.INET);
+                if (boundSocket.get() != null) {
+                    serverChannel.bind(boundSocket.get());
+                } else {
+                    InetSocketAddress socketAddress = new InetSocketAddress(localhost, portNumber);
+                    serverChannel.bind(socketAddress);
+
+                    boundSocket.set((InetSocketAddress) serverChannel.getLocalAddress());
+
+                    BoundTransportAddress boundAddress = boundAddress();
+                    for (BoundAddressListener listener : boundAddressListeners) {
+                        listener.addressBound(boundAddress);
+                    }
+                }
+            } catch (Exception e) {
+                throw new BindTransportException("Failed to bind to " + NetworkAddress.format(localhost, portNumber), e);
+            }
+
+            return null;
+        });
+
+        return serverChannel;
+    }
+
+    @Override
+    protected void doStart() {
+        // Mark the service as active, we'll start the listener when ES is ready
+        this.active = true;
+    }
+
+    // package private for testing
+    synchronized void startListener() {
+        assert enabled(environment);
+
+        if (this.serverChannel != null || this.active == false) {
+            return;
+        }
+
+        this.serverChannel = setupSocket();
+        this.listenerThreadLatch = new CountDownLatch(1);
+
+        new Thread(() -> {
+            assert serverChannel != null && listenerThreadLatch != null;
+            while (serverChannel.isOpen()) {
+                AccessController.doPrivileged((PrivilegedAction<Void>) () -> {
+                    try (SocketChannel channel = serverChannel.accept()) {} catch (IOException e) {
+                        logger.debug("encountered exception while responding to readiness check request", e);
+                    } catch (Exception other) {
+                        logger.warn("encountered unknown exception while responding to readiness check request", other);
+                    }
+                    return null;
+                });
+            }
+            listenerThreadLatch.countDown();
```

Comment:
The use of a latch here is cute, maybe move the `countDown` to a finally block.  Alternatively, the latch could be dropped in favour of retaining a reference to the Thread, and using `join`, which I think it the same semantic as what is achieved with the latch.

Reply:
Good idea on making it part of a finally. I prefer the latch instead of the join because it makes the stop code cleaner. Depending of how startup is interrupted stop() can get called without start being called, which means I'd have to check if the thread reference is not null and then only await. Adds one more branch I need to test and look into, with the latch on the other hand, if it was never set to 1 it will just continue.

Reply:
Ok.